h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -67,14 +69,18 @@
  * @see #startGridSearch(Key, Key, HyperSpaceWalker, Recovery, int)
  */
 public final class GridSearch<MP extends Model.Parameters> {
+  private static final int DEFAULT_MAX_CONSECUTIVE_MODEL_FAILURES = 10;
+  
   public final Key<Grid> _result;
   public final Job<Grid> _job;
   public final Recovery<Grid> _recovery;
   public final int _parallelism;
+  public int _maxConsecutiveModelFailures = DEFAULT_MAX_CONSECUTIVE_MODEL_FAILURES;
 
   /** Walks hyper space and for each point produces model parameters. It is
    *  used only locally to fire new model builders.  */
   private final transient HyperSpaceWalker<MP, ?> _hyperSpaceWalker;
+  private final AtomicInteger _consecutiveModelFailures = new AtomicInteger();
```

Comment:
Do we want this to be a property of the grid search itself? I think it is conceptually a feature of the walker - the walker already receives notifications about failed models and it could be its responsibility to say there are no more models to train because nothing we tried succeeded (hasNext method).

Also the walker can be injected from outside of GS and any behavior can thus be introduced without GS knowing.

Reply:
> I think it is conceptually a feature of the walker

imo, the responsibility of the walker is limited to providing new model builders: I agree that the walker can stop the walking but it can not fail the job (or it would be in a very unusual way, like throwing on `hasNext()`), and I think that interrupting abruptly the entire job is a responsibility of the entity that owns the job, ie. GS.
Another way to look at it: if a model fails, there could be a policy to retry the training. Who would be responsible for this, the walker or the grid search? Imo, it's the search that would be responsible for this, as retrying is outside the scope of iterating. So, in general, I would keep any decision outside the iteration logic per se outside the walker.

Reply:
@sebhrusen do you need all this functionality in grid search? 

For example - can it be limited to eg. 10 consecutive failed models with no model trained? This is something you can control from the walker and it is still useful, failures can be handled on AutoML side.

To be honest I don't feel comfortable reviewing changes in GS of this kind. If we could find a way to make it useful in AutoML without structural changes in GS then that would be ideal but if the scope has to stay this broad I think we will have to hold off for after the major release.

Reply:
@michalkurka I updated grid to fail on consecutive failures iff no model has been trained yet. Is it ok for you?
I could also change the default `_maxConsecutiveModelFailures` to 0 for grids, and set it to 10 only in `AutoML` but this would mean that users can still pass wrong hyperparams and have a grid running forever without producing anything.

Reply:
I think that looks good

I would keep the default "unlimited", we can then introduce a parameter for grid search to expose it to users and pick a reasonable default